lib/features/push_notification/presentation/websocket/web_socket_queue_handler.dart

```diff
@@ -0,0 +1,122 @@
+
+import 'dart:async';
+import 'dart:collection';
+
+import 'package:core/utils/app_logger.dart';
+import 'package:flutter/cupertino.dart';
+import 'package:tmail_ui_user/features/push_notification/presentation/websocket/web_socket_message.dart';
+
+typedef ProcessMessageCallback = Future<void> Function(WebSocketMessage message);
+typedef OnErrorCallback = void Function(dynamic error, StackTrace stackTrace);
+
+class WebSocketQueueHandler {
+  static const int _maxQueueSize = 128;
+  static const int _maxProcessedIdsSize = 128;
+
+  final Queue<WebSocketMessage> _messageQueue = Queue<WebSocketMessage>();
+  final Queue<String> _processedMessageIds = Queue<String>();
+
+  Completer<void>? _processingLock;
+
+  final _queueController = StreamController<WebSocketMessage>.broadcast();
+
+  final ProcessMessageCallback processMessageCallback;
+  final OnErrorCallback? onErrorCallback;
+
+  WebSocketQueueHandler({
+    required this.processMessageCallback,
+    this.onErrorCallback,
+  }) {
+    _queueController.stream.listen((_) {
+      _processQueue();
+    });
+  }
+
+  void enqueue(WebSocketMessage message) {
+    if (isMessageProcessed(message.id)) {
+      log('WebSocketQueueHandler::enqueue:Message ${message.id} already processed, skipping');
+      return;
+    }
+
+    if (queueSize >= _maxQueueSize) {
+      log('WebSocketQueueHandler::enqueue:Queue full, removing oldest message');
+      _messageQueue.removeFirst();
+    }
+
+    _messageQueue.add(message);
+    _queueController.add(message);
+  }
+
+  Future<void> _processQueue() async {
+    if (_processingLock != null) {
+      return;
+    }
+
+    _processingLock = Completer<void>();
+
+    try {
+      while (queueSize > 0) {
+        final message = _messageQueue.first;
+
+        try {
+          await processMessageCallback(message);
+          _addToProcessedMessages(message.id);
+          _messageQueue.removeFirst();
```

Comment:
why not move this to `finally` block?

Reply:
should check size before remove to avoid throw exception

Reply:
this `_messageQueue.removeFirst();` maybe remove the item which not process  because we also remove inside `processMessageCallback`.

So, i think we should replace L59 with `_messageQueue.removeFirst();`

Reply:
Oke updated